## AI-Driven Financial Trend Analysis of Microsoft, Tesla, and Apple (2022–2024)

In [3]:
import pandas as pd

In [5]:
df = pd.read_csv('Financial_data.csv')
df.head()

Company   Fiscal Year Total Revenue($) Net Income($) Total Assets($)  \
0  Microsoft          2022         1,98,270        72,738        3,64,840   
1  Microsoft          2023         2,11,915        72,361        4,11,395   
2  Microsoft          2024         2,44,996        88,205        5,12,000   
3      Tesla          2022           81,460        12,556          82,338   
4      Tesla          2023           96,773        14,997        1,06,620   

  Total Liabilities($) Operating Cash Flow($)  
0             1,98,298                 89,034  
1             2,11,206                 95,022  
2             2,45,000               1,10,000  
3               36,440                 14,724  
4               44,000                 16,000

In [6]:
# Remove commas and convert to numeric
columns_to_clean = ['Total Revenue($)', 'Net Income($)', 'Total Assets($)', 'Total Liabilities($)', 'Operating Cash Flow($)']
for col in columns_to_clean:
    df[col] = df[col].replace(',', '', regex=True).astype(int)


In [7]:
df.head()

Company   Fiscal Year  Total Revenue($)  Net Income($)  Total Assets($)  \
0  Microsoft          2022            198270          72738           364840   
1  Microsoft          2023            211915          72361           411395   
2  Microsoft          2024            244996          88205           512000   
3      Tesla          2022             81460          12556            82338   
4      Tesla          2023             96773          14997           106620   

   Total Liabilities($)  Operating Cash Flow($)  
0                198298                   89034  
1                211206                   95022  
2                245000                  110000  
3                 36440                   14724  
4                 44000                   16000

In [9]:

df.columns = df.columns.str.strip()


### Year-over-Year Financial Growth Analysis

In [13]:
df = df.sort_values(['Company', 'Fiscal Year'])

df['Revenue Growth (%)'] = df.groupby('Company')['Total Revenue($)'].pct_change() * 100
df['Net Income Growth (%)'] = df.groupby('Company')['Net Income($)'].pct_change() * 100
df['Assets Growth (%)'] = df.groupby('Company')['Total Assets($)'].pct_change() * 100
df['Liabilities Growth (%)'] = df.groupby('Company')['Total Liabilities($)'].pct_change() * 100
df['Cash Flow Growth (%)'] = df.groupby('Company')['Operating Cash Flow($)'].pct_change() * 100


In [14]:
df

Company  Fiscal Year  Total Revenue($)  Net Income($)  Total Assets($)  \
6      Apple         2022            394328          99803           352755   
7      Apple         2023            383285          96995           352583   
8      Apple         2024            400000         105000           360000   
0  Microsoft         2022            198270          72738           364840   
1  Microsoft         2023            211915          72361           411395   
2  Microsoft         2024            244996          88205           512000   
3      Tesla         2022             81460          12556            82338   
4      Tesla         2023             96773          14997           106620   
5      Tesla         2024             97683           7000           122000   

   Total Liabilities($)  Operating Cash Flow($)  Revenue Growth (%)  \
6                302083                  122151                 NaN   
7                290437                  110543           -2.800461   
8                310000                  115000            4.360985   
0                198298                   89034                 NaN   
1                211206                   95022            6.882030   
2                245000                  110000           15.610504   
3                 36440                   14724                 NaN   
4                 44000                   16000           18.798183   
5                 50000                   12000            0.940345   

   Net Income Growth (%)  Assets Growth (%)  Liabilities Growth (%)  \
6                    NaN                NaN                     NaN   
7              -2.813543          -0.048759               -3.855232   
8               8.253003           2.103618                6.735712   
0                    NaN                NaN                     NaN   
1              -0.518299          12.760388                6.509395   
2              21.895773          24.454600               16.000492   
3                    NaN                NaN                     NaN   
4              19.440905          29.490636               20.746432   
5             -53.323998          14.425061               13.636364   

   Cash Flow Growth (%)  
6                   NaN  
7             -9.502992  
8              4.031915  
0                   NaN  
1              6.725521  
2             15.762665  
3                   NaN  
4              8.666123  
5            -25.000000

## Conclusion: 
### Microsoft: 
####            Revenue grew steadily, with a significant 15.6% jump in 2024, showing strong business expansion.
####            Net income rose sharply in 2024 by ~21.9%, indicating improved profitability.
####            Both assets and liabilities increased, suggesting active investments or expansions.

###   Tesla:  
####            Tesla’s revenue grew slightly in 2024 (0.94%), showing stagnation.
####            Net income fell by 53.3% in 2024 — a major decline in profitability, possibly due to rising costs or lower margins.
####            Cash flow dropped by 25%, another red flag.

###    Apple:  
####            Apple’s revenue dipped in 2023 (-2.8%) but bounced back in 2024 (4.36%).
####            Net income followed the same trend — dip then recovery.
####            Stable growth in assets and cash flow, indicating financial health.



## Net Income Margin (%) - Profitability Measure

In [17]:
# Calculate Net Income as a % of Revenue (Profit Margin)
df['Net Income Margin (%)'] = (df['Net Income($)'] / df['Total Revenue($)']) * 100

df[['Company', 'Fiscal Year', 'Net Income Margin (%)']]


Company  Fiscal Year  Net Income Margin (%)
6      Apple         2022              25.309641
7      Apple         2023              25.306234
8      Apple         2024              26.250000
0  Microsoft         2022              36.686337
1  Microsoft         2023              34.146238
2  Microsoft         2024              36.002629
3      Tesla         2022              15.413700
4      Tesla         2023              15.497091
5      Tesla         2024               7.166037

##  Conclusion: 

#### Microsoft shows a robust profit margin that remains stable or increases, showcasing strong profitability.

#### Apple maintains a consistent profit margin with slight growth, indicating efficiency and stable cost management.

#### Tesla’s profitability has declined, especially in 2024, indicating potential challenges in cost control, competition, or market conditions.

## Debt to Asset Ratio - Financial Risk Measure

In [18]:
# Calculate how much of assets are financed by debt
df['Debt to Asset Ratio'] = df['Total Liabilities($)'] / df['Total Assets($)']

df[['Company', 'Fiscal Year', 'Debt to Asset Ratio']]


Company  Fiscal Year  Debt to Asset Ratio
6      Apple         2022             0.856354
7      Apple         2023             0.823741
8      Apple         2024             0.861111
0  Microsoft         2022             0.543520
1  Microsoft         2023             0.513390
2  Microsoft         2024             0.478516
3      Tesla         2022             0.442566
4      Tesla         2023             0.412681
5      Tesla         2024             0.409836

## Conclusion: 

#### Apple has a relatively high debt-to-asset ratio, with values consistently above 0.8, indicating that a large portion of its assets is financed by debt.

#### Microsoft has a much lower debt-to-asset ratio, which is decreasing over the years, suggesting a trend towards reducing reliance on debt for financing assets.

#### Tesla also has a moderate ratio, but similar to Microsoft, its debt reliance is gradually decreasing.

## Average Operating Cash Flow Growth

In [20]:
# Average Operating Cash Flow Growth by Company
avg_cash_flow_growth = df.groupby('Company')['Cash Flow Growth (%)'].mean()
print(avg_cash_flow_growth)


Company
Apple        -2.735539
Microsoft    11.244093
Tesla        -8.166938
Name: Cash Flow Growth (%), dtype: float64


## Conclusion: 

####  Apple has a negative average cash flow growth of -2.74%, indicating a decline in its operating cash flow growth over the years.

####  Microsoft has a positive average growth of 11.24%, suggesting consistent improvement in its operating cash flow.

#### Tesla also shows a negative average growth of -8.17%, pointing to a decrease in its operating cash flow over time.